cmd/vodloadtester/vodloadtester.go

```diff
@@ -0,0 +1,423 @@
+	}
+
+	if cliFlags.DirectUpload || cliFlags.ResumableUpload {
+		if fileName, err = utils.GetFile(cliFlags.Filename, strings.ReplaceAll(hostName, ".", "_")); err != nil {
+			if err == utils.ErrNotFound {
+				glog.Fatalf("file %s not found\n", cliFlags.Filename)
+			} else {
+				glog.Fatalf("error getting file %s: %v\n", cliFlags.Filename, err)
+			}
+		}
+
+		if cliFlags.DirectUpload {
+			glog.Infof("Launching direct upload load test for %q", fileName)
+			vt.directUploadLoadTest(fileName, runnerInfo)
+		}
+		if cliFlags.ResumableUpload {
+			glog.Infof("Launching resumable upload load test for %q", fileName)
+			vt.resumableUploadLoadTest(fileName, runnerInfo)
+		}
+	}
+
+}
+
+func (vt *vodLoadTester) requestUploadUrls(assetName string) (*api.UploadUrls, error) {
+	uploadUrls, err := vt.lapi.RequestUpload(assetName)
+	if err != nil {
+		return nil, err
+	}
+	return uploadUrls, nil
+}
+
+func (vt *vodLoadTester) directUploadLoadTest(fileName string, runnerInfo string) {
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < int(vt.cliFlags.VideoAmount); i += int(vt.cliFlags.Simultaneous) {
+		for j := 0; j < int(vt.cliFlags.Simultaneous); j++ {
+			fmt.Printf("Uploading video %d/%d\n", i+j+1, vt.cliFlags.VideoAmount)
+			wg.Add(1)
+			go func() {
+				uploadTest := uploadTest{
+					StartTime:  time.Now(),
+					RunnerInfo: runnerInfo,
+					Kind:       "directUpload",
+				}
+				rndAssetName := fmt.Sprintf("load_test_direct_%s", randName())
+				requestedUpload, err := vt.requestUploadUrls(rndAssetName)
+				defer vt.lapi.DeleteAsset(requestedUpload.Asset.ID)
+
+				if err != nil {
+					glog.Errorf("Error requesting upload urls: %v", err)
+					uploadTest.RequestUploadSuccess = 0
+					uploadTest.ErrorMessage = err.Error()
+					vt.writeResultNdjson(uploadTest)
+					return
+				} else {
+					uploadTest.RequestUploadSuccess = 1
+					uploadTest.AssetID = requestedUpload.Asset.ID
+					uploadTest.TaskID = requestedUpload.Task.ID
+				}
+
+				err = vt.uploadAsset(fileName, requestedUpload.Url)
+
+				if err != nil {
+					glog.Errorf("Error uploading asset: %v", err)
+					uploadTest.UploadSuccess = 0
+				} else {
+					uploadTest.UploadSuccess = 1
+					if vt.cliFlags.TaskCheck {
+						err := vt.checkTaskProcessing(5*time.Second, api.TaskOnlyId{ID: uploadTest.TaskID})
+						if err != nil {
+							uploadTest.TaskCheckSuccess = 0
+							uploadTest.ErrorMessage = err.Error()
+						} else {
+							uploadTest.TaskCheckSuccess = 1
+						}
+					}
+
+				}
+				vt.writeResultNdjson(uploadTest)
+				wg.Done()
+			}()
+		}
+		time.Sleep(vt.cliFlags.StartDelayDuration)
+	}
+
+	wg.Wait()
+}
+
+func (vt *vodLoadTester) resumableUploadLoadTest(fileName, runnerInfo string) {
+	wg := &sync.WaitGroup{}
+
+	for i := 0; i < int(vt.cliFlags.VideoAmount); i += int(vt.cliFlags.Simultaneous) {
+		for j := 0; j < int(vt.cliFlags.Simultaneous); j++ {
+			fmt.Printf("rUpload video %d/%d\n", i+j+1, vt.cliFlags.VideoAmount)
+			wg.Add(1)
+			go func() {
```

Comment:
I think that with this logic you are actually doing `VideoAmount * Simultaneous` concurrent uploads. What you want is probably to spawn `Simultaneous` routines and iterate (together) over `VideoAmount` videos, which you can do using a channel. We can pair on this if you want, but this might also help: https://go.dev/blog/pipelines

Reply:
Sounds good, although I think something is missing here. With the previous version, I could specify a `StartDelayDuration` between groups, so that I could fire groups of video to be imported separated by a specified delay. 

This way instead, subsequent videos are uploaded only on task completion, am I wrong?

Reply:
For example, I cannot make batches of 5 videos to be uploaded every 30 seconds